backend/hitas/services/email.py

```diff
@@ -0,0 +1,243 @@
+    get_first_sale_purchase_date,
+    get_first_sale_purchase_price,
+    get_latest_sale_purchase_date,
+    get_latest_sale_purchase_price,
+    prefetch_latest_sale,
+)
+from hitas.services.condition_of_sale import condition_of_sale_queryset
+from hitas.services.indices import (
+    subquery_apartment_current_surface_area_price,
+    subquery_apartment_first_sale_acquisition_price_index_adjusted,
+)
+from hitas.services.thirty_year_regulation import get_thirty_year_regulation_results_for_housing_company
+from hitas.utils import monthify
+from hitas.views.apartment import ApartmentDetailSerializer
+from hitas.views.utils.pdf import render_to_pdf
+from users.models import User
+
+
+def send_confirmed_max_price_calculation_email(
+    calculation_id: UUID,
+    template_name: str,
+    recipients: list[str],
+    user: User,
+) -> None:
+    confirmed_max_price_calculation = get_hitas_object_or_404(ApartmentMaximumPriceCalculation, uuid=calculation_id)
+    filename, pdf = render_confirmed_max_price_calculation_pdf(confirmed_max_price_calculation, user)
+
+    template = get_hitas_object_or_404(
+        EmailTemplate,
+        name=template_name,
+        type=EmailTemplateType.CONFIRMED_MAX_PRICE_CALCULATION,
+    )
+    send_pdf_via_email(body=template.text, recipients=recipients, filename=filename, pdf=pdf)
+
+
+def render_confirmed_max_price_calculation_pdf(
+    confirmed_max_price_calculation: ApartmentMaximumPriceCalculation,
+    user: User,
+) -> tuple[str, bytes]:
+    pdf_body = get_hitas_object_or_404(PDFBody, name=PDFBodyName.CONFIRMED_MAX_PRICE_CALCULATION)
+    filename = f"Enimmäishintalaskelma {confirmed_max_price_calculation.apartment.address}.pdf"
+    context = {
+        "maximum_price_calculation": confirmed_max_price_calculation,
+        "user": user,
+        "body_parts": pdf_body.texts,
+    }
+    pdf = render_to_pdf(template="confirmed_maximum_price.jinja", context=context)
+    return filename, pdf
+
+
+def send_unconfirmed_max_price_calculation_email(
+    apartment_id: UUID,
+    calculation_date: datetime.date,
+    additional_info: str,
+    template_name: str,
+    recipients: list[str],
+    user: User,
+) -> None:
+    filename, pdf = render_unconfirmed_max_price_calculation_pdf(
+        apartment_id=apartment_id,
+        calculation_date=calculation_date,
+        additional_info=additional_info,
+        user=user,
+    )
+
+    template = get_hitas_object_or_404(
+        EmailTemplate,
+        name=template_name,
+        type=EmailTemplateType.UNCONFIRMED_MAX_PRICE_CALCULATION,
+    )
+    send_pdf_via_email(body=template.text, recipients=recipients, filename=filename, pdf=pdf)
+
+
+def render_unconfirmed_max_price_calculation_pdf(
+    apartment_id: UUID,
+    calculation_date: datetime.date,
+    additional_info: str,
+    user: User,
+) -> tuple[str, bytes]:
+    calculation_month = monthify(calculation_date)
+
+    apartment: Optional[ApartmentWithAnnotations] = (
```

Comment:
I don't like the amount of duplication **here** vs `ApartmentViewSet` 🙈

Reply:
Yes, I know. 😫 The issue here was just that refactoring and simplifying the queryset and PDF template is a lot of work, since there isn't a good interface in place (it's reusing the apartment endpoint interface, which has a lot of unnecessary data, which is also formatted in a custom way). Maybe it's still worth doing, even with the limited time.